sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs

```diff
@@ -0,0 +1,146 @@
+    {
+        internal const string DiagnosticId = "S4050";
+        private const string MessageFormat = "Provide an implementation for: {0}.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static class MethodName
+        {
+            public const string OperatorPlus = "operator+";
+            public const string OperatorMinus = "operator-";
+            public const string OperatorEquals = "operator==";
+            public const string OperatorNotEquals = "operator!=";
+
+            public const string ObjectEquals = "Object.Equals";
+            public const string ObjectGetHashCode = "Object.GetHashCode";
+        }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                if (classSymbol == null ||
+                    classDeclaration.Identifier.IsMissing ||
+                    !IsPubliclyAccessible(classSymbol))
+                {
+                    return;
+                }
+
+                var missingMethods = FindMissingMethods(classSymbol);
+                if (missingMethods.Any())
+                {
+                    string missingMethodsMessage = DiagnosticReportHelper.CreateStringFromArgs(missingMethods);
+                    c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(), missingMethodsMessage));
+                }
+            },
+            SyntaxKind.ClassDeclaration);
+        }
+
+        private static bool IsPubliclyAccessible(ISymbol symbol)
+        {
+            var effectiveAccessibility = symbol.GetEffectiveAccessibility();
+
+            return effectiveAccessibility == Accessibility.Public ||
+                effectiveAccessibility == Accessibility.Protected ||
+                effectiveAccessibility == Accessibility.ProtectedOrInternal;
+        }
+
+        private static IEnumerable<string> FindMissingMethods(INamedTypeSymbol classSymbol)
+        {
+            var implementedMethods = GetImplementedMethods(classSymbol).ToHashSet();
+            var requiredMethods = new HashSet<string>();
+
+            if (implementedMethods.Contains(MethodName.OperatorPlus))
+            {
+                requiredMethods.Add(MethodName.OperatorMinus);
+                requiredMethods.Add(MethodName.OperatorEquals);
+                requiredMethods.Add(MethodName.OperatorNotEquals);
+                requiredMethods.Add(MethodName.ObjectEquals);
```

Comment:
That wouldn't work. The methods are required when at least one of the operators is present. If we added it by default, every class would need to implement Equals and GetHashCode. This might cause a couple of FPs ;)